ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Query.hs

```diff
@@ -146,14 +198,27 @@ queryDecodeNodeToClient ::
 queryDecodeNodeToClient codecConfig queryVersion blockVersion
   = case queryVersion of
       TopLevelQueryDisabled -> decodeBlockQuery
-      QueryVersion1 -> do
+      QueryVersion1         -> handleTopLevelQuery
+      QueryVersion2         -> handleTopLevelQuery
+  where
+    handleTopLevelQuery :: Decoder s (SomeSecond Query blk)
+    handleTopLevelQuery = do
         size <- decodeListLen
         tag  <- decodeWord8
         case (size, tag) of
-          (2, 0) -> decodeBlockQuery
-          (1, 1) -> return (SomeSecond GetSystemStart)
+          (2, 0) -> requireVersion "BlockQuery"           QueryVersion1   decodeBlockQuery
+          (1, 1) -> requireVersion "GetSystemStart"       QueryVersion1 $ return (SomeSecond GetSystemStart)
+          (1, 2) -> requireVersion "GetChainBlockNo"      QueryVersion2 $ return (SomeSecond GetChainBlockNo)
+          (1, 3) -> requireVersion "GetChainPoint"        QueryVersion2 $ return (SomeSecond GetChainPoint)
           _      -> fail $ "Query: invalid size and tag" <> show (size, tag)
-  where
+
+    requireVersion :: String -> QueryVersion -> Decoder s (SomeSecond Query blk) -> Decoder s (SomeSecond Query blk)
+    requireVersion name expectedVersion f =
```

Comment:
I prefer if `name` were somehow derived from `f` (which would requires rejigging `f`).

Or instead, replace `name` with the `(size, tag)`. My concern is that human error will lead to `name` not matching `f`.

Reply:
I've eliminated the `name` argument.